### ISSUE TITLE
一起讨论下CRNN样本作成策略问题

### ISSUE THREAD

**Issue by moondaiy** [chineseocr/chinese-ocr#70]
楼主好，目前正在进行CRNN识别，针对场景例如发票等，我目前所做的实验如下:

(1). 准备网络上下载的各种XX新闻.txt文件作为语料文件，放入courps中的(并没有考虑文字平衡问题)
(2). 准备目标预测背景图片作为背景候选图片
大概作成了200万个图片训练集样本和20万个验证集样本，使用pytorch版本进行GPU训练，并加载作者提供的预训练模型netCRNN.pth。
目前结论是在测试集可以得到89%准确率 Loss < 0.1 

但是实际在预测我们目标图片时 CRNN识别还是蛮低的，尤其对于数字和符号(类似“()”这样的)，基本上没有对的情况。

所以和大家一起讨论下，CRNN样本作成策略问题。感觉怎么样平衡才算合理。。比如是否需要考虑加入单独数字组合，专有名词等。训练过程中是否长度和动态调整，不是每次都拿10个进行。。

二. 另外想请问下我们自带的CRNN中的预测模型 model_ac97.pth是谁训练出来的额？

另外还有一个很奇怪的地方。就是我使用GPU训练的模型 放到CPU版本上需要更改代码。
    #model.load_state_dict(torch.load(path))
    model.load_state_dict(torch.load(path, map_location=lambda storage, loc: storage))
model_ac97.pth 就不用。。。。不知道大家有没有遇到过。。

**Comment by wenlihaoyu** [comment]
需要看你构造的样本是否存在“()”这样的字符，贡献出来的模型（这个模型我自己训练出来的）是支持识别的。预训练模型netCRNN.pth只训练了两个字符。
识别结果：
<img width="782" alt="12" src="https://user-images.githubusercontent.com/7800267/35213585-15cf07aa-ff99-11e7-9c10-bb1f413f3652.png">
图片文本检测结果：
![13](https://user-images.githubusercontent.com/7800267/35213649-48f644ea-ff99-11e7-95e1-e5df299aee3f.jpg)

**Comment by wenlihaoyu** [comment]
你也可以放一些英文语料在里面

**Comment by favpdf** [comment]
@moondaiy 你训练的时候，字符集用的是keys.py定义好的，还是根据预料重新构建的？训练过程中，有没有先训练短字符串然后再逐渐训练长字符串？比如，先训练2个字符，4个字符，最后训练10个字符。

**Comment by moondaiy** [comment]
@favpdf 是使用keys.py定义中的字符，训练过程中没有进行短字符串的(10个字以下)的训练，直接使用10个字符形成一个串的样本进行训练(但是是预加载了LZ提供的与训练模型，LZ在上文说这个模型是2个字的模型)。
所以我个人感觉在训练过程中，应该会有一些策略，比如学习率调整等问题。包括您所说的 字符个数问题。

**Comment by favpdf** [comment]
@moondaiy 我试过调整学习率，效果并不理想。当前的0.005应该是作者根据他经验给出的较优值吧。我最近也在搞OCR，方便发个联系方式给我邮箱吗？咱们讨论一下？我的邮箱：favpdf@gmail.com

**Comment by wenlihaoyu** [comment]
参数设置为0.005，就算不加载预训练模型，直接训练十个字符的模型，模型的准确率也能够达到95%。主要还是靠训练集的构造，调参数只能从起很少的作用。最早closed的几个issues也介绍过。

**Comment by moondaiy** [comment]
@wenlihaoyu 请问下，我在看程序源码时候，貌似是可以进行不定长样本训练的，就是可以不用10个字符构成一个样本集 ，比如英文单词 hello 单独作为一个样本，请问这样训练也是没问题的吧？因为做成一次样本 比较话费时间，所有还是想先确定好。再进行

**Comment by huang-aoyun** [comment]
如果不finetune,训练数据集（可能几百万吧）大概要迭代多少次？

**Comment by wenlihaoyu** [comment]
你注意样本平衡，100多万就够了，不平衡的话，就不知道了，没有测试过。

**Comment by huang-aoyun** [comment]
100多万的数据集大概迭代多少次呢？

**Comment by wenlihaoyu** [comment]
要看收敛了，一般跑10多个小时就行（具体依赖你电脑的配置及数据集质量）。多少次没有关心。

**Comment by hustbobo** [comment]
@moondaiy 你目前有考虑语料平衡的问题么？有没有好的策略可以保证训练数据的平衡性问题

**Comment by moondaiy** [comment]
@hustbobo  最新的训练是考虑到平衡语料的问题了. 但是平衡的单位是字符. 先无条件的制作10000个字符大小的训练集,统计当前训练集中每个字符出现的频率 ,完后再继续加入别的训练样本字符,如果该字符大于平均频率的5倍(英语字符为10倍) 则放弃该训练文本的加入...然后继续添加

不过这个办法还是会出现不平衡问题...正在训练ing 看看效果吧

**Comment by hustbobo** [comment]
@moondaiy 目前也是这样考虑，但是有些字语料中出现的太少，字符也是以N个字符同时出现的，我也在尝试这种做法效果行不行，另外，你的语料库是用的搜狗那个新闻的，还是有更好的语料库了，望推荐，哈哈

**Comment by moondaiy** [comment]
怎么说呢.语料就开源出来的就哪些...xxx新闻什么的 .这个也是做中文NLP的一个短板

**Comment by houzi09** [comment]
@moondaiy  请问训练的时候使用不用字数长度的样本，效果怎么样？我看代码，感觉如果长度不够最大长度，后面就是0，这样是不是就相当于是alphabet字典里的第一个字符 ' ，这样训练会不会有问题？

**Comment by wenlihaoyu** [comment]
@houzi09 不够长的用空白字符填充，不是0

**Comment by houzi09** [comment]
@wenlihaoyu 是指空格吗？请问这个操作的代码是在哪个位置呢？这样会不会影响训练效果？
如果是用空格来代替，相当于把图片里的空白部分当成空格，每个图里的空白部分长度不一的时候，会不会导致训练有问题？

**Comment by wenlihaoyu** [comment]
你说的这个问题不存在，因为计算准确率的时候会trim掉空格，pytorch使用的'-'替代（utils.py文件），做空格填充，主要是英文，当时训练中文是10个字符，英文是一个单词，所以就用了填充

**Comment by houzi09** [comment]
@wenlihaoyu  不好意思，没找到trim掉空格的操作。
用'-'替代不会有问题吗？我看 alphabet 里有'-'，那不会当做正常值吗？
